### ISSUE TITLE
Use of Helm Template for JupyterHub

### ISSUE THREAD

**Issue by zmkhazi** [jupyterhub/jupyterhub#2188]
I am new to Kubernetes deployments. The aim is to install JupyterHub on K8s cluster. Followed the instructions in zero to Jupyterhub. It talks about helm install. Here in my office we are not using tiller . Hence can execute helm install, but  was told to use helm template to manually get the helm charts  and push on the K8S cluster. 

I was able to get the helm client installed. Proceeded to the next step in creating the chart  for JupyterHub. Tried with the stable and development release as well. Get the following errors. Any help on this is appreciated. 

**Tried with the Stable Release**
C:\Helm\windows-amd64>helm template C:\Helm\JupyterHub\jupyterhub-0.7.0\jupyterhub
Error: render error in "jupyterhub/templates/proxy/deployment.yaml": template: jupyterhub/templates/proxy/deployment.yaml:26:32: executing "jupyterhub/templates/proxy/deployment.yaml" at <include (print $.Tem...>: error calling include: template: jupyterhub/templates/hub/secret.yaml:9:19: executing "jupyterhub/templates/hub/secret.yaml" at <required "Proxy toke...>: error calling required: Proxy token must be a 32 byte random string generated with `openssl rand -hex 32`! 

**Tried with the Developer’s release as well.**
C:\Helm\windows-amd64>helm template C:\Helm\JupyterHub\jupyterhub-0.8-dc14a29\jupyterhub
Error: render error in "jupyterhub/templates/proxy/deployment.yaml": template: jupyterhub/templates/proxy/deployment.yaml:26:32: executing "jupyterhub/templates/proxy/deployment.yaml" at <include (print $.Tem...>: error calling include: template: jupyterhub/templates/hub/secret.yaml:9:19: executing "jupyterhub/templates/hub/secret.yaml" at <required "Proxy toke...>: error calling required: Proxy token must be a 32 byte random string generated with `openssl rand -hex 32`!

Thanks

**Comment by consideRatio** [comment]
The Helm chart will render the Templates based on values you provide, if you dont, you lack important things and the generation of k8s resources from the templates complains. This is what happens.

Create a config.yaml as z2jh documentation describes, then pass it to the `helm template` command with the `--values` command line option.

**Comment by zmkhazi** [comment]
Thanks for the info. Will look into it.

**Comment by zmkhazi** [comment]
> The Helm chart will render the Templates based on values you provide, if you dont, you lack important things and the generation of k8s resources from the templates complains. This is what happens.
> Create a config.yaml as z2jh documentation describes, then pass it to the helm template command with the --values command line option.
> # write this to a file and save it as config.yaml for example
> proxy:
>   # only the proxy's secretToken is required
>   secretToken: <generate this using "openssl rand -hex 32">
> # run this later, generate the jupyterhub chart based on provided values in your config.yaml
> helm template jupyterhub/jupyterhub --values config.yaml

Here is what I did to get the template  without the tiller. See the detailed Log below. I need direction on what to next.  The aim to get Jupyterhub installed on a Kubernetes Cluster.  I don’t think I have to run the helm upgrade as specified in step 2  as my gut feeling is it needs tiller and here in my company we don't use tiller ?

https://zero-to-jupyterhub.readthedocs.io/en/stable/setup-jupyterhub.html

" STEPS FOLLOWED"
Downloaded the  JupterHub Helm chart and unzipped the tar file and used this as a reference point in the helm template. Later ran the helm template command.

[root]# helm template jupyterhub --values config.yaml

"Results from the Log"

[root]# helm template jupyterhub --values config.yaml
---
# Source: jupyterhub/templates/hub/secret.yaml
kind: Secret
apiVersion: v1
metadata:
  name: hub-secret
  labels:
    component: hub
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: Tiller
type: Opaque
data:
  proxy.token: "NjFiMzFjNzE1ZDc2MDMzOWMxZWUzMzNjNzYzNGNiZDg3MDQzY2RjMGE4Mzk5ZGJmNTFkMTBjYjFiY2ZlMGI1OA=="
  hub.cookie-secret: "ZjA0NzUxYjBhMDdmODkyNzJjNWI4MzgzOWEyZjNmMTM1Nzg4YzY1YjJmMDM5NmY3NzA0ZjEyMTM0MjVjNzNiZg=="

---
# Source: jupyterhub/templates/hub/configmap.yaml
kind: ConfigMap
apiVersion: v1
metadata:
  name: hub-config
  labels:
    component: hub
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: Tiller
data:
  cull.enabled: "true"
  cull.users: "false"
  cull.timeout: "3600"
  cull.every: "600"
  cull.concurrency: "10"
  cull.max-age: "0"
  auth.type: "dummy"
  auth.state.enabled: "false"
  auth.admin.access: "true"
  singleuser.network-tools.image.name: "jupyterhub/k8s-network-tools"
  singleuser.network-tools.image.tag: "0.7.0"
  singleuser.cloud-metadata: |
    enabled: false
    ip: 169.254.169.254
  singleuser.start-timeout: "300"
  singleuser.image-pull-policy: "IfNotPresent"
  singleuser.cmd: "jupyterhub-singleuser"
  singleuser.events: "true"
  singleuser.uid: "1000"
  singleuser.fs-gid: "100"
  singleuser.node-selector: "{}"
  singleuser.storage.type: "dynamic"
  singleuser.storage.home_mount_path: "/home/jovyan"
  singleuser.storage.extra-volumes: "[]"
  singleuser.storage.extra-volume-mounts: "[]"
  singleuser.storage.capacity: "10Gi"
  singleuser.storage.dynamic.pvc-name-template: "claim-{username}{servername}"
  singleuser.storage.dynamic.volume-name-template: "volume-{username}{servername}"
  singleuser.storage.dynamic.storage-access-modes: "[ReadWriteOnce]"
  singleuser.memory.guarantee: "1G"
  singleuser.extra-labels: |
    hub.jupyter.org/network-access-hub: "true"
  kubespawner.common-labels: |
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: jupyterhub
  hub.allow-named-servers: "false"
  hub.concurrent-spawn-limit: "64"
  hub.consecutive-failure-limit: "5"
  hub.base_url: "/"
  hub.db_url: "sqlite:///jupyterhub.sqlite"
  token.cookie_secret: "f04751b0a07f89272c5b83839a2f3f135788c65b2f0396f7704f1213425c73bf"
  debug.enabled: "false"

---
# Source: jupyterhub/templates/hub/pvc.yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: hub-db-dir
  labels:
    component: hub
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: Tiller
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: "1Gi"

---
# Source: jupyterhub/templates/hub/rbac.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: hub
  labels:
    component: hub
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: Tiller
---
kind: Role
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
  name: hub
  labels:
    component: hub
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: Tiller
rules:
  - apiGroups: [""]       # "" indicates the core API group
    resources: ["pods", "persistentvolumeclaims"]
    verbs: ["get", "watch", "list", "create", "delete"]
  - apiGroups: [""]       # "" indicates the core API group
    resources: ["events"]
    verbs: ["get", "watch", "list"]
---
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
  name: hub
  labels:
    component: hub
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: Tiller
subjects:
  - kind: ServiceAccount
    name: hub
    namespace: default
roleRef:
  kind: Role
  name: hub
  apiGroup: rbac.authorization.k8s.io

---
# Source: jupyterhub/templates/image-puller/rbac.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: hook-image-awaiter
  labels:
    component: image-puller
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: Tiller
    hub.jupyter.org/deletable: "true"
  annotations:
    "helm.sh/hook": pre-install,pre-upgrade
    "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded
    "helm.sh/hook-weight": "0"
---
kind: Role
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
  name: hook-image-awaiter
  labels:
    component: image-puller
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: Tiller
    hub.jupyter.org/deletable: "true"
  annotations:
    "helm.sh/hook": pre-install,pre-upgrade
    "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded
    "helm.sh/hook-weight": "0"
rules:
  - apiGroups: ["apps"]       # "" indicates the core API group
    resources: ["daemonsets"]
    verbs: ["get"]
---
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
  name: hook-image-awaiter
  labels:
    component: image-puller
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: Tiller
    hub.jupyter.org/deletable: "true"
  annotations:
    "helm.sh/hook": pre-install,pre-upgrade
    "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded
    "helm.sh/hook-weight": "0"
subjects:
  - kind: ServiceAccount
    name: hook-image-awaiter
    namespace: default
roleRef:
  kind: Role
  name: hook-image-awaiter
  apiGroup: rbac.authorization.k8s.io

---
# Source: jupyterhub/templates/hub/service.yaml
apiVersion: v1
kind: Service
metadata:
  name: hub
  labels:
    component: hub
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: Tiller
spec:
  type: ClusterIP
  selector:
    component: hub
    app: jupyterhub
    release: RELEASE-NAME
  ports:
    - protocol: TCP
      port: 8081
      targetPort: 8081

---
# Source: jupyterhub/templates/proxy/service.yaml
apiVersion: v1
kind: Service
metadata:
  name: proxy-api
  labels:
    component: proxy-api
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: Tiller
spec:
  selector:
    component: proxy
    app: jupyterhub
    release: RELEASE-NAME
  ports:
    - protocol: TCP
      port: 8001
      targetPort: 8001
---
apiVersion: v1
kind: Service
metadata:
  name: proxy-public
  labels:
    component: proxy-public
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: Tiller
spec:
  selector:
    # TODO: Refactor to utilize the helpers
    component: proxy
    release: RELEASE-NAME
  ports:
    - name: http
      port: 80
      protocol: TCP
      targetPort: 8000
      # allow proxy.service.nodePort for http
  type: LoadBalancer

---
# Source: jupyterhub/templates/image-puller/daemonset.yaml
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  name: hook-image-puller
  labels:
    component: hook-image-puller
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: Tiller
    hub.jupyter.org/deletable: "true"
  annotations:
    "helm.sh/hook": pre-install,pre-upgrade
    "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded
    "helm.sh/hook-weight": "-10"
spec:
  selector:
    matchLabels:
      component: hook-image-puller
      app: jupyterhub
      release: RELEASE-NAME
  updateStrategy:
    type: RollingUpdate
    rollingUpdate:
      maxUnavailable: 100%
  template:
    metadata:
      labels:
        component: hook-image-puller
        app: jupyterhub
        release: RELEASE-NAME
    spec:
      terminationGracePeriodSeconds: 0
      automountServiceAccountToken: false
      initContainers:
        - name: image-pull-singleuser
          image: jupyterhub/k8s-singleuser-sample:0.7.0
          imagePullPolicy: IfNotPresent
          command:
            - /bin/sh
            - -c
            - echo "Pulling complete"
        - name: image-pull-metadata-block
          image: jupyterhub/k8s-network-tools:0.7.0
          imagePullPolicy: IfNotPresent
          command:
            - /bin/sh
            - -c
            - echo "Pulling complete"
      nodeSelector: {}
      containers:
        - name: pause
          image: gcr.io/google_containers/pause:3.0
---

---
# Source: jupyterhub/templates/hub/deployment.yaml
apiVersion: apps/v1beta2
kind: Deployment
metadata:
  name: hub
  labels:
    component: hub
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: Tiller
spec:
  replicas: 1
  selector:
    matchLabels:
      component: hub
      app: jupyterhub
      release: RELEASE-NAME
  strategy:
    rollingUpdate: null
    type: Recreate
  template:
    metadata:
      labels:
        component: hub
        app: jupyterhub
        release: RELEASE-NAME
        hub.jupyter.org/network-access-proxy-api: "true"
        hub.jupyter.org/network-access-proxy-http: "true"
        hub.jupyter.org/network-access-singleuser: "true"
      annotations:
        # This lets us autorestart when the secret changes!
        checksum/config-map: 00b721f4c50b2dadee36fcafaf9b10f45843a3310dfb3f371c3222334faa096c
        checksum/secret: e089f3e337aa067011a60977d9906d9b46e35e9f9735cc1f1c179be3cc373c4c
        prometheus.io/path: /hub/metrics
        prometheus.io/scrape: "true"
    spec:
      nodeSelector: {}
      affinity:
        podAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
            - weight: 1
              podAffinityTerm:
                topologyKey: kubernetes.io/hostname
                labelSelector:
                  matchExpressions:
                    - key: component
                      operator: In
                      values: ['proxy']
                    - key: release
                      operator: In
                      values: ["RELEASE-NAME"]
      volumes:
        - name: config
          configMap:
            name: hub-config
        - name: secret
          secret:
            secretName: hub-secret
        - name: hub-db-dir
          persistentVolumeClaim:
            claimName: hub-db-dir
      serviceAccountName: hub
      securityContext:
        runAsUser: 1000
        fsGroup: 1000
      containers:
        - name: hub
          image: jupyterhub/k8s-hub:0.7.0
          command:
            - jupyterhub
            - --config
            - /srv/jupyterhub_config.py
            - --upgrade-db
          volumeMounts:
            - mountPath: /etc/jupyterhub/config/
              name: config
            - mountPath: /etc/jupyterhub/secret/
              name: secret
            - mountPath: /srv/jupyterhub
              name: hub-db-dir
          resources:
            requests:
              cpu: 200m
              memory: 512Mi
          imagePullPolicy: IfNotPresent
          env:
            - name: SINGLEUSER_IMAGE
              value: "jupyterhub/k8s-singleuser-sample:0.7.0"
            - name: JPY_COOKIE_SECRET
              valueFrom:
                secretKeyRef:
                  name: hub-secret
                  key: hub.cookie-secret
            - name: POD_NAMESPACE
              valueFrom:
                fieldRef:
                  fieldPath: metadata.namespace
            - name: CONFIGPROXY_AUTH_TOKEN
              valueFrom:
                secretKeyRef:
                  name: hub-secret
                  key: proxy.token
          ports:
            - containerPort: 8081
              name: hub

---
# Source: jupyterhub/templates/proxy/deployment.yaml
apiVersion: apps/v1beta2
kind: Deployment
metadata:
  name: proxy
  labels:
    component: proxy
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: Tiller
spec:
  replicas: 1
  selector:
    matchLabels:
      component: proxy
      app: jupyterhub
      release: RELEASE-NAME
  template:
    metadata:
      labels:
        component: proxy
        app: jupyterhub
        release: RELEASE-NAME
        hub.jupyter.org/network-access-hub: "true"
        hub.jupyter.org/network-access-singleuser: "true"
      annotations:
        # This lets us autorestart when the secret changes!
        checksum/hub-secret: f01dbdc7ed6dfdccf2fa2fba7a78c287c89d5f3488ab50d738f3036e75bcfea1
        checksum/proxy-secret: 01ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546b
    spec:
      nodeSelector: {}
      terminationGracePeriodSeconds: 60
      affinity:
        podAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
            - weight: 1
              podAffinityTerm:
                topologyKey: kubernetes.io/hostname
                labelSelector:
                  matchExpressions:
                    - key: component
                      operator: In
                      values: ['hub']
                    - key: release
                      operator: In
                      values: ["RELEASE-NAME"]
      containers:
        - name: chp
          image: jupyterhub/configurable-http-proxy:3.0.0
          command:
            - configurable-http-proxy
            - --ip=0.0.0.0
            - --api-ip=0.0.0.0
            - --api-port=8001
            - --default-target=http://$(HUB_SERVICE_HOST):$(HUB_SERVICE_PORT)
            - --error-target=http://$(HUB_SERVICE_HOST):$(HUB_SERVICE_PORT)/hub/error
            - --port=8000
          resources:
            requests:
              cpu: 200m
              memory: 512Mi
          env:
            - name: CONFIGPROXY_AUTH_TOKEN
              valueFrom:
                secretKeyRef:
                  name: hub-secret
                  key: proxy.token
          imagePullPolicy: IfNotPresent
          ports:
            - containerPort: 8000
              name: proxy-public
            - containerPort: 8001
              name: api

---
# Source: jupyterhub/templates/image-puller/job.yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: hook-image-awaiter
  labels:
    component: image-puller
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: Tiller
    hub.jupyter.org/deletable: "true"
  annotations:
    "helm.sh/hook": pre-install,pre-upgrade
    "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded
    "helm.sh/hook-weight": "10"
spec:
  template:
    metadata:
      labels:
        component: image-puller
        app: jupyterhub
        release: RELEASE-NAME
    spec:
      restartPolicy: Never
      serviceAccountName: hook-image-awaiter
      containers:
        - image: jupyterhub/k8s-image-awaiter:0.7.0
          name: hook-image-awaiter
          imagePullPolicy: IfNotPresent
          command:
            - /image-awaiter
            - -ca-path=/var/run/secrets/kubernetes.io/serviceaccount/ca.crt
            - -auth-token-path=/var/run/secrets/kubernetes.io/serviceaccount/token
            - -api-server-address=https://$(KUBERNETES_SERVICE_HOST):$(KUBERNETES_SERVICE_PORT)
            - -namespace=default
            - -daemonset=hook-image-puller

---
# Source: jupyterhub/templates/hub/pdb.yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
metadata:
  name: hub
  labels:
    component: hub
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: Tiller
spec:
  minAvailable: 1
  selector:
    matchLabels:
      component: hub
      app: jupyterhub
      release: RELEASE-NAME

---
# Source: jupyterhub/templates/proxy/pdb.yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
metadata:
  name: proxy
  labels:
    component: proxy
    app: jupyterhub
    release: RELEASE-NAME
    chart: jupyterhub-0.7.0
    heritage: Tiller
spec:
  minAvailable: 1
  selector:
    matchLabels:
      component: proxy
      app: jupyterhub
      release: RELEASE-NAME
---
# Source: jupyterhub/templates/hub/netpol.yaml
---
# Source: jupyterhub/templates/ingress.yaml
---
# Source: jupyterhub/templates/pod-culler/deployment.yaml
---
# Source: jupyterhub/templates/pod-culler/rbac.yaml
---
# Source: jupyterhub/templates/proxy/autohttps/configmap-nginx.yaml
---
# Source: jupyterhub/templates/proxy/autohttps/deployment.yaml
---
# Source: jupyterhub/templates/proxy/autohttps/ingress-internal.yaml
---
# Source: jupyterhub/templates/proxy/autohttps/rbac.yaml
---
# Source: jupyterhub/templates/proxy/autohttps/service.yaml
---
# Source: jupyterhub/templates/proxy/netpol.yaml
---
# Source: jupyterhub/templates/proxy/secret.yaml
---
# Source: jupyterhub/templates/singleuser/image-credentials-secret.yaml
---
# Source: jupyterhub/templates/singleuser/netpol.yaml

**Comment by zmkhazi** [comment]
> Read about the helm template command with
> 
> helm template --help
> 
> 
> 
> Find out how to output this into multiple files
> 
> 
> Use kubectl to install things
> 
> 
> 
> This is not something i would recommend, this is the absolute last resort, you may very well run into a lot of other issues that I don't think you will find help from someone to sort out as it will be a bit too complicated to debug.

Sir, thanks for your guidance. I will work on those lines you specified.
 what is your best recommendation on this ? Do you think having tiller will ease the complexity ?

**Comment by consideRatio** [comment]
You are welcome @zmkhazi =)

Yepp using the Helm chart without tiller seems to me like using a car without any motor. It would be possible to get the car moving without it, but it has lost a lot of its purpose. I recommend installing tiller on the cluster, and using helm to deploy the chart, by following the guide at https://z2jh.jupyter.org

**Comment by zmkhazi** [comment]
> Read about the helm template command with
> 
> helm template --help
> 
> 
> 
> Find out how to output this into multiple files
> 
> 
> Use kubectl to install things
> 
> 
> 
> This is not something i would recommend, this is the absolute last resort, you may very well run into a lot of other issues that I don't think you will find help from someone to sort out as it will be a bit too complicated to debug.
I was able to out put the template into multiple files under outdir. Need to execute the kubectl commands. Need the right scoop here.

While I wait for my IT partners to get the Tiller installed on the K8S cluster, they have asked me to try the kubectl command.  Getting tiller approved might take some time, as I have been told that we can't get tiller at this point of time due to the lack of role based access control in that configuration.

 I am not clear which kubectl commands, I need to execute. I looked into multiple site, they all have references to tiller. All I want is to execute the kubectl commands without tiller. 

Here is the outdir.
[root@/data/helm/outdir/jupyterhub/templates]# ll
total 6
drwxr-x--- 2 root root 203 Oct  6 00:27 hub
drwxr-x--- 2 root root  85 Oct  6 00:27 image-puller
drwxr-x--- 2 root root  89 Oct  6 00:27 proxy
